Add StandardLoggingPayload to DynamoDB Logging

marty-sullivan opened the pull request:
## Add StandardLoggingPayload to DynamoDB Logging

DynamoDB logging integration was not yet updated to have the StandardLoggingPayload and so it was inconsistent with other logging integrations. Notably, it was not logging completions for all model call types, only text completions.

## Type

🐛 Bug Fix
🧹 Refactoring

## Changes

- Implemented the StandardLoggingPayload for the DynamoDB logging integration
- This allows for all model types to be logged to DynamoDB rather than just chat completions, and now has 

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locally

I tested in my development environment and I did not get errors and the DynamoDB target gets logging info for all model types and in the StandardLoggingPayload format.

More testing may be needed as the input needs to be sanitized for DynamoDB insertion so it's possible that my sanitization method is not taking into account all possibilities

marty-sullivan committed:
try standardloggingpayload in dynamodb

krrishdholakia commented:
Hi @marty-sullivan can you please add any unit tests + run the checks noted here - https://docs.litellm.ai/docs/extras/contributing_code

marty-sullivan commented:
@krrishdholakia I ran the `make test-unit` and everything passed except redis related ones, presumably just because I don't have redis configured locally

```
rootdir: /Users/mjs472/repos/litellm
configfile: pyproject.toml
plugins: asyncio-0.25.3, anyio-4.5.2, mock-3.14.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None
collected 84 items                                                                                                                                                                                  

tests/litellm/caching/test_redis_cache.py FF                                                                                                                                                  [  2%]
tests/litellm/caching/test_redis_cluster_cache.py ..                                                                                                                                          [  4%]
tests/litellm/integrations/test_athina.py .....                                                                                                                                               [ 10%]
tests/litellm/litellm_core_utils/test_dd_tracing.py ....                                                                                                                                      [ 15%]
tests/litellm/litellm_core_utils/test_safe_json_dumps.py .........                                                                                                                            [ 26%]
tests/litellm/litellm_core_utils/test_streaming_handler.py ....                                                                                                                               [ 30%]
tests/litellm/llms/azure_ai/chat/test_azure_ai_transformation.py ..                                                                                                                           [ 33%]
tests/litellm/llms/bedrock/test_base_aws_llm.py ..                                                                                                                                            [ 35%]
tests/litellm/llms/custom_httpx/test_handle_audio_file.py ...                                                                                                                                 [ 39%]
tests/litellm/proxy/common_utils/test_http_parsing_utils.py ...                                                                                                                               [ 42%]
tests/litellm/proxy/db/test_check_migration.py .                                                                                                                                              [ 44%]
tests/litellm/proxy/db/test_prisma_client.py .                                                                                                                                                [ 45%]
tests/litellm/proxy/hooks/test_proxy_track_cost_callback.py .                                                                                                                                 [ 46%]
tests/litellm/proxy/pass_through_endpoints/test_llm_pass_through_endpoints.py ...s                                                                                                            [ 51%]
tests/litellm/proxy/spend_tracking/test_spend_management_endpoints.py .....                                                                                                                   [ 57%]
tests/litellm/proxy/test_caching_routes.py .......                                                                                                                                            [ 65%]
tests/litellm/proxy/test_proxy_cli.py ........                                                                                                                                                [ 75%]
tests/litellm/proxy/test_route_llm_request.py ........                                                                                                                                        [ 84%]
tests/litellm/rerank_api/test_rerank_main.py .                                                                                                                                                [ 85%]
tests/litellm/test_main.py ...........                                                                                                                                                        [ 98%]
tests/litellm/test_model_prices_and_context_window_schema.py .                                                                                                                                [100%]

============================================================================================= FAILURES ==============================================================================================
______________________________________________________________________________ test_redis_cache_async_increment[None] _______________________________________________________________________________

namespace = None

    @pytest.mark.parametrize("namespace", [None, "test"])
    @pytest.mark.asyncio
    async def test_redis_cache_async_increment(namespace):
>       redis_cache = RedisCache(namespace=namespace)

tests/litellm/caching/test_redis_cache.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
litellm/caching/redis_cache.py:82: in __init__
    self.redis_client = get_redis_client(**redis_kwargs)
litellm/_redis.py:249: in get_redis_client
    redis_kwargs = _get_redis_client_logic(**env_overrides)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

env_overrides = {}, redis_kwargs = {}, _startup_nodes = None, _sentinel_nodes = None, _sentinel_password = None

    def _get_redis_client_logic(**env_overrides):
        """
        Common functionality across sync + async redis client implementations
        """
        ### check if "os.environ/<key-name>" passed in
        for k, v in env_overrides.items():
            if isinstance(v, str) and v.startswith("os.environ/"):
                v = v.replace("os.environ/", "")
                value = get_secret(v)  # type: ignore
                env_overrides[k] = value
    
        redis_kwargs = {
            **_redis_kwargs_from_environment(),
            **env_overrides,
        }
    
        _startup_nodes: Optional[Union[str, list]] = redis_kwargs.get("startup_nodes", None) or get_secret(  # type: ignore
            "REDIS_CLUSTER_NODES"
        )
    
        if _startup_nodes is not None and isinstance(_startup_nodes, str):
            redis_kwargs["startup_nodes"] = json.loads(_startup_nodes)
    
        _sentinel_nodes: Optional[Union[str, list]] = redis_kwargs.get("sentinel_nodes", None) or get_secret(  # type: ignore
            "REDIS_SENTINEL_NODES"
        )
    
        if _sentinel_nodes is not None and isinstance(_sentinel_nodes, str):
            redis_kwargs["sentinel_nodes"] = json.loads(_sentinel_nodes)
    
        _sentinel_password: Optional[str] = redis_kwargs.get(
            "sentinel_password", None
        ) or get_secret_str("REDIS_SENTINEL_PASSWORD")
    
        if _sentinel_password is not None:
            redis_kwargs["sentinel_password"] = _sentinel_password
    
        _service_name: Optional[str] = redis_kwargs.get("service_name", None) or get_secret(  # type: ignore
            "REDIS_SERVICE_NAME"
        )
    
        if _service_name is not None:
            redis_kwargs["service_name"] = _service_name
    
        if "url" in redis_kwargs and redis_kwargs["url"] is not None:
            redis_kwargs.pop("host", None)
            redis_kwargs.pop("port", None)
            redis_kwargs.pop("db", None)
            redis_kwargs.pop("password", None)
        elif "startup_nodes" in redis_kwargs and redis_kwargs["startup_nodes"] is not None:
            pass
        elif (
            "sentinel_nodes" in redis_kwargs and redis_kwargs["sentinel_nodes"] is not None
        ):
            pass
        elif "host" not in redis_kwargs or redis_kwargs["host"] is None:
>           raise ValueError("Either 'host' or 'url' must be specified for redis.")
E           ValueError: Either 'host' or 'url' must be specified for redis.

litellm/_redis.py:169: ValueError
______________________________________________________________________________ test_redis_cache_async_increment[test] _______________________________________________________________________________

namespace = 'test'

    @pytest.mark.parametrize("namespace", [None, "test"])
    @pytest.mark.asyncio
    async def test_redis_cache_async_increment(namespace):
>       redis_cache = RedisCache(namespace=namespace)

tests/litellm/caching/test_redis_cache.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
litellm/caching/redis_cache.py:82: in __init__
    self.redis_client = get_redis_client(**redis_kwargs)
litellm/_redis.py:249: in get_redis_client
    redis_kwargs = _get_redis_client_logic(**env_overrides)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

env_overrides = {}, redis_kwargs = {}, _startup_nodes = None, _sentinel_nodes = None, _sentinel_password = None

    def _get_redis_client_logic(**env_overrides):
        """
        Common functionality across sync + async redis client implementations
        """
        ### check if "os.environ/<key-name>" passed in
        for k, v in env_overrides.items():
            if isinstance(v, str) and v.startswith("os.environ/"):
                v = v.replace("os.environ/", "")
                value = get_secret(v)  # type: ignore
                env_overrides[k] = value
    
        redis_kwargs = {
            **_redis_kwargs_from_environment(),
            **env_overrides,
        }
    
        _startup_nodes: Optional[Union[str, list]] = redis_kwargs.get("startup_nodes", None) or get_secret(  # type: ignore
            "REDIS_CLUSTER_NODES"
        )
    
        if _startup_nodes is not None and isinstance(_startup_nodes, str):
            redis_kwargs["startup_nodes"] = json.loads(_startup_nodes)
    
        _sentinel_nodes: Optional[Union[str, list]] = redis_kwargs.get("sentinel_nodes", None) or get_secret(  # type: ignore
            "REDIS_SENTINEL_NODES"
        )
    
        if _sentinel_nodes is not None and isinstance(_sentinel_nodes, str):
            redis_kwargs["sentinel_nodes"] = json.loads(_sentinel_nodes)
    
        _sentinel_password: Optional[str] = redis_kwargs.get(
            "sentinel_password", None
        ) or get_secret_str("REDIS_SENTINEL_PASSWORD")
    
        if _sentinel_password is not None:
            redis_kwargs["sentinel_password"] = _sentinel_password
    
        _service_name: Optional[str] = redis_kwargs.get("service_name", None) or get_secret(  # type: ignore
            "REDIS_SERVICE_NAME"
        )
    
        if _service_name is not None:
            redis_kwargs["service_name"] = _service_name
    
        if "url" in redis_kwargs and redis_kwargs["url"] is not None:
            redis_kwargs.pop("host", None)
            redis_kwargs.pop("port", None)
            redis_kwargs.pop("db", None)
            redis_kwargs.pop("password", None)
        elif "startup_nodes" in redis_kwargs and redis_kwargs["startup_nodes"] is not None:
            pass
        elif (
            "sentinel_nodes" in redis_kwargs and redis_kwargs["sentinel_nodes"] is not None
        ):
            pass
        elif "host" not in redis_kwargs or redis_kwargs["host"] is None:
>           raise ValueError("Either 'host' or 'url' must be specified for redis.")
E           ValueError: Either 'host' or 'url' must be specified for redis.

litellm/_redis.py:169: ValueError
========================================================================================= warnings summary ==========================================================================================
../../micromamba/envs/litellmdev/lib/python3.12/site-packages/pydantic/_internal/_config.py:295
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/site-packages/pydantic/_internal/_config.py:295: PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.10/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)

litellm/utils.py:169
  /Users/mjs472/repos/litellm/litellm/utils.py:169: DeprecationWarning: open_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with resources.open_text(

tests/litellm/llms/azure_ai/chat/test_azure_ai_transformation.py::test_azure_ai_request_format[True]
  /Users/mjs472/repos/litellm/litellm/llms/openai/openai.py:276: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
    _response_headers=dict(raw_response.headers),
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

tests/litellm/proxy/db/test_prisma_client.py::test_should_update_prisma_schema
  /Users/mjs472/repos/litellm/tests/litellm/proxy/db/test_prisma_client.py:19: PytestWarning: Value of environment variable DISABLE_SCHEMA_UPDATE type should be str, but got None (type: NoneType); converted to str implicitly
    monkeypatch.setenv("DISABLE_SCHEMA_UPDATE", None)

tests/litellm/proxy/db/test_prisma_client.py::test_should_update_prisma_schema
  /Users/mjs472/repos/litellm/tests/litellm/proxy/db/test_prisma_client.py:31: PytestWarning: Value of environment variable DISABLE_SCHEMA_UPDATE type should be str, but got None (type: NoneType); converted to str implicitly
    monkeypatch.setenv("DISABLE_SCHEMA_UPDATE", None)

tests/litellm/proxy/db/test_prisma_client.py::test_should_update_prisma_schema
  /Users/mjs472/repos/litellm/tests/litellm/proxy/db/test_prisma_client.py:34: PytestWarning: Value of environment variable DISABLE_SCHEMA_UPDATE type should be str, but got None (type: NoneType); converted to str implicitly
    monkeypatch.setenv("DISABLE_SCHEMA_UPDATE", None)  # Set env var opposite to param

tests/litellm/proxy/pass_through_endpoints/test_llm_pass_through_endpoints.py::TestBaseOpenAIPassThroughHandler::test_base_openai_pass_through_handler
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/site-packages/_pytest/python.py:148: PytestUnhandledCoroutineWarning: async def functions are not natively supported and have been skipped.
  You need to install a suitable plugin for your async framework, for example:
    - anyio
    - pytest-asyncio
    - pytest-tornasync
    - pytest-trio
    - pytest-twisted
    warnings.warn(PytestUnhandledCoroutineWarning(msg.format(nodeid)))

tests/litellm/rerank_api/test_rerank_main.py::test_rerank_infer_region_from_model_arn
tests/litellm/test_main.py::test_url_with_format_param[True-bedrock/anthropic.claude-3-5-sonnet-20240620-v1:0]
tests/litellm/test_main.py::test_url_with_format_param[True-bedrock/invoke/anthropic.claude-3-5-sonnet-20240620-v1:0]
tests/litellm/test_main.py::test_url_with_format_param[False-bedrock/anthropic.claude-3-5-sonnet-20240620-v1:0]
tests/litellm/test_main.py::test_url_with_format_param[False-bedrock/invoke/anthropic.claude-3-5-sonnet-20240620-v1:0]
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/site-packages/botocore/auth.py:425: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    datetime_now = datetime.datetime.utcnow()

tests/litellm/test_main.py::test_url_with_format_param[True-bedrock/invoke/anthropic.claude-3-5-sonnet-20240620-v1:0]
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/site-packages/pydantic/main.py:426: UserWarning: Pydantic serializer warnings:
    Expected `str` but got `MagicMock` with value `<MagicMock name='mock().j...em__()' id='5349993568'>` - serialized value may not be as expected
    return self.__pydantic_serializer__.to_python(

tests/litellm/test_main.py::test_url_with_format_param[True-bedrock/invoke/anthropic.claude-3-5-sonnet-20240620-v1:0]
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/unittest/mock.py:2566: RuntimeWarning: coroutine 'serve' was never awaited
    return tuple.__new__(cls, (name, args, kwargs))
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

tests/litellm/test_main.py::test_url_with_format_param[True-anthropic/claude-3-5-sonnet]
  /Users/mjs472/micromamba/envs/litellmdev/lib/python3.12/site-packages/pydantic/main.py:426: UserWarning: Pydantic serializer warnings:
    Expected `str` but got `MagicMock` with value `<MagicMock name='mock().j...em__()' id='5342707248'>` - serialized value may not be as expected
    return self.__pydantic_serializer__.to_python(

tests/litellm/test_main.py::test_url_with_format_param_openai[False-gpt-4o-mini]
  /Users/mjs472/repos/litellm/litellm/llms/openai/openai.py:427: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
    headers = dict(raw_response.headers)
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
====================================================================================== short test summary info ======================================================================================
FAILED tests/litellm/caching/test_redis_cache.py::test_redis_cache_async_increment[None] - ValueError: Either 'host' or 'url' must be specified for redis.
FAILED tests/litellm/caching/test_redis_cache.py::test_redis_cache_async_increment[test] - ValueError: Either 'host' or 'url' must be specified for redis.
======================================================================= 2 failed, 81 passed, 1 skipped, 16 warnings in 8.85s ========================================================================
make: *** [test-unit] Error 1
```

marty-sullivan committed:
Merge branch 'BerriAI:main' into dynamodb
